From: Bernhard Kaindl <bernhard.kaindl@cloud.com>
Date: Tue, 22 Jul 2025 12:00:00 +0200
Subject: Track vcpu->nonaffine_time for each vCPU

- Add a new time counter, nonaffine_time to struct vcpu.
- Add a new inline function to check if a vCPU is/was nonaffine.
- Use it in the vcpu_runstate_change handler for vcpu->nonaffine_time

Note:
- We cannot change struct vcpu_runstate_info to add this counter:
  It is part of the guest runstate shared memory area which is
  in the guest ABI that is frozen and cannot be changed.

Update vcpu_runstate_get() to return a snapshot of the current
non-affine vCPU time at the time of call.

Signed-off-by: Bernhard Kaindl <bernhard.kaindl@cloud.com>

diff --git a/xen/common/sched/core.c b/xen/common/sched/core.c
index b5a3213e3a6e..e88244c675e7 100644
--- a/xen/common/sched/core.c
+++ b/xen/common/sched/core.c
@@ -260,6 +260,17 @@ static inline void vcpu_urgent_count_update(struct vcpu *v)
     }
 }
 
+/*
+ * For accounting purposes, we need to know whether a vCPU is (or was just)
+ * running outside of the soft affinity mask of the scheduling unit.
+ */
+static inline bool is_running_nonaffine(const struct vcpu *v,
+                                        const struct sched_unit *unit)
+{
+    return v->runstate.state == RUNSTATE_running && has_soft_affinity(unit) &&
+           !cpumask_test_cpu(v->processor, unit->cpu_soft_affinity);
+}
+
 static inline void vcpu_runstate_change(
     struct vcpu *v, int new_state, s_time_t new_entry_time)
 {
@@ -285,9 +296,15 @@ static inline void vcpu_runstate_change(
     /* Serialization: ->schedule_lock (see ASSERT() above). */
     with_seq_write(&v->runstate_seq)
     {
+        v->is_running_nonaffine = is_running_nonaffine(v, unit);
+
         if ( delta > 0 )
         {
             v->runstate.time[v->runstate.state] += delta;
+
+            if ( v->is_running_nonaffine )
+                v->runstate_extra.nonaffine_time += delta;
+
             v->runstate.state_entry_time = new_entry_time;
         }
 
@@ -309,21 +326,41 @@ void sched_guest_idle(void (*idle) (void), unsigned int cpu)
     atomic_dec(&per_cpu(sched_urgent_count, cpu));
 }
 
-void vcpu_runstate_get(const struct vcpu *v,
-                       struct vcpu_runstate_info *runstate)
+/**
+ * vcpu_runstate_get(): Return vCPU time spent in different runstates
+ *
+ * @param v:        vCPU to get runstate times (since vCPU start)
+ * @param runstate: Return time spent in each runstate.
+ *                  This structure is part of the runstate memory areas
+ *                  shared with the domains which is part of the ABI
+ *                  with domains that is frozen and cannot be changed.
+ *                  To return additional values, use e.g. the return
+ *                  value(no need to change all callers) of this function.
+ * @returns         struct with non-affine running time since vcpu creation
+ */
+struct vcpu_runstate_extra vcpu_runstate_get(
+    const struct vcpu *v, struct vcpu_runstate_info *runstate)
 {
     struct seqcount seq = SEQCNT_ZERO();
     const struct seqcount *s = likely(v == current) ? &seq : &v->runstate_seq;
+    struct vcpu_runstate_extra ret;
 
     until_seq_read(s)
     {
         s_time_t delta;
 
         *runstate = v->runstate;
+        ret = v->runstate_extra;
         delta = NOW() - runstate->state_entry_time;
         if ( delta > 0 )
+        {
             runstate->time[runstate->state] += delta;
+            if ( v->is_running_nonaffine )
+                ret.nonaffine_time += delta;
+        }
     }
+
+    return ret;
 }
 
 uint64_t vcpu_runstate_get_running(const struct vcpu *v)
diff --git a/xen/include/xen/sched.h b/xen/include/xen/sched.h
index 612cdbf35db4..6dd416d5954b 100644
--- a/xen/include/xen/sched.h
+++ b/xen/include/xen/sched.h
@@ -214,6 +214,13 @@ struct vcpu
     struct seqcount  runstate_seq;
     unsigned int     new_state;
 
+    /* Accessed under runstate_seq */
+    struct vcpu_runstate_extra {
+        uint64_t nonaffine_time; /* Time running outside soft_affinity mask */
+    } runstate_extra;
+    /* Signal whether the vCPU is running on a non-affine CPU */
+    bool is_running_nonaffine;
+
     /* Has the FPU been initialised? */
     bool             fpu_initialised;
     /* Has the FPU been used since it was last saved? */
@@ -1125,8 +1132,8 @@ int vcpu_set_hard_affinity(struct vcpu *v, const cpumask_t *affinity);
 int vcpu_affinity_domctl(struct domain *d, uint32_t cmd,
                          struct xen_domctl_vcpuaffinity *vcpuaff);
 
-void vcpu_runstate_get(const struct vcpu *v,
-                       struct vcpu_runstate_info *runstate);
+struct vcpu_runstate_extra vcpu_runstate_get(
+    const struct vcpu *v, struct vcpu_runstate_info *runstate);
 uint64_t vcpu_runstate_get_running(const struct vcpu *v);
 uint64_t get_cpu_idle_time(unsigned int cpu);
 void sched_guest_idle(void (*idle) (void), unsigned int cpu);
